R/db.R

```diff
@@ -40,6 +40,47 @@ setMethod(
   }
 )
 
+# Snowflake --------------------------------------------------------------------
+
+# Simple class prototype to avoid messages about unknown classes from setMethod
+setClass("Snowflake", where = class_cache)
+
+#' @rdname odbcConnectionTables
+#' @details If `exact` is set to TRUE, then escape underscores in the catalog, schema,
+#' and table names.  There is empirical evidence that this leads to performance improvements.
+setMethod(
+  "odbcConnectionTables",
+  c("Snowflake", "character"),
+  function(conn, name, catalog_name = NULL, schema_name = NULL, table_type = NULL, exact = FALSE) {
+
+    if ( !exact ) return(callNextMethod())
+
+    if(!is.null(catalog_name)) catalog_name <- gsub("([^\\\\])(_)", "\\1\\\\_", catalog_name)
```

Comment:
Based on 

> ODBC 2.x drivers do not support search patterns in the CatalogName argument in SQLTables. ODBC 3*.x* drivers accept search patterns in this argument if the SQL_ATTR_ODBC_VERSION environment attribute is set to SQL_OV_ODBC3; they do not accept search patterns in this argument if it is set to SQL_OV_ODBC2.

I wonder if we should also be setting the `SQL_ATTR_ODBC_VERSION` env var? I have no sense of whether drivers are commonly v2 or v3.

Reply:
Hey - thanks for giving this some thought.  I thought about putting it in the generic but opted to defer since:

* It's not clear it is needed; thought we can always introduce it for back-ends where it makes a difference. I am still a bit surprised it makes a difference for `Snowflake` TBH, but at least we have empirical evidence there.  I could be off here, but my base case is that with these catalog methods the greatest gains to be had are by using fully qualified identifiers - for example when looking at the column structure, querying for `[catalog].[schema].[name]`.
* Not sure what if `\` is a uniform escape character across back-ends and drivers.
* Haven't had a chance to test it elsewhere, and to note if there are any drivers that may be misbehaving.

At any rate, easy enough to move to the generic - let me know if you would rather do that.

In terms of API version, [we assume](https://github.com/r-dbi/odbc/blob/main/src/Makevars.in#L4) it is 3.x.

Reply:
Given that the odbc spec uses `_` for pattern matching, I think it makes sense to implement it in the generic. IMO this is confusing wrinkle of odbc that we should make sure to surface in the docs as much as possible. 

I suspect all databases will use `\`, but if we discover later that one doesn't we can retrieve and cache `SQL_SEARCH_PATTERN_ESCAPE` in the connection object.

Reply:
We just had an internal report of slowness on databricks, and given the presence of `_`, I bet this is the root cause.

Reply:
Great / i hope escaping underscores will help there as well.  Should have an opportunity to address the second round of feedback in the next couple of days.

Reply:
I'll also take a stab at a slightly different approach where we escape by default, but allow `I()` to revert to the previous behaviour. I think that is likely to be what 99% of people want, but still gives us an escape hatch if they really did want pattern matching.

Reply:
Thanks for all the thoughtful feedback!

Hope I took everything into account with the latest iteration.  Let me know if I missed anything.